src/servers/Http1Server.cc

```diff
@@ -338,7 +338,7 @@ Http::One::Server::writeControlMsgAndCall(HttpReply *rep, AsyncCall::Pointer &ca
     // paranoid: ContentLengthInterpreter has cleaned non-generated replies
     rep->removeIrrelevantContentLength();
     rep->header.putStr(Http::HdrType::CONNECTION, "keep-alive");
-    httpHdrMangleList(&rep->header, http->request, http->al, ROR_REPLY);
+    httpHdrMangleList(&rep->header, http->request, nullptr, http->al, ROR_REPLY);
```

Comment:
Why are we not passing the response here? For example, would the following work?

```suggestion
    httpHdrMangleList(&rep->header, http->request, rep, http->al, ROR_REPLY);
```

Reply:
> I don't know what is the effect of passing reply object to httpHdrMangleList() in Http1Server.cc.

The effect of passing the reply object to httpHdrMangleList() should not depend on the caller. If you have reasons to think it depends on the caller, let's discuss. Otherwise, you _do_ know what the effect is.

> for this, I am afraid it changing will cause problem...

I understand your concern. It is always tempting to fix the exact problem you are suffering from and minimize the change risk by keeping the rest of the code intact. I hope you understand my concern as well: When this tempting approach is used in the majority of cases, it creates unmanageable code and causes exponentially more bugs long-term.

In this particular case, we should keep in mind that _not_ changing this caller (i.e. using nullptr for the newly added parameter) will cause (different) problems and will also preserve (in this caller context) a problem the PR is fixing. There is no universal formula for deciding when to preserve a bug in context A while fixing it in context B, but, in this particular case, the reasons for fixing it seem to outweigh the reasons for preserve it.

Your analysis may reach a different conclusion, and I am happy to discuss the problems I have missed, but they should be more substantial than the general fear of breaking something.

> I am not sure how to test this code path.

Send an HTTP 100 Continue response to Squid under test?

> Can I just apply your change?

It is your call -- there are too many variables here for me to answer that question correctly. If you can test this change, please do so.

Reply:
OK...it may be obvious to you that "there should be no problem", but I am not familar with the internal logic of squid. Both code paths are something "give you the reply, and based on the reply, modify the reply itself", there could be pointer issues like changing the destination actually changes what you are reading, which breaks linked lists, or so. Or there could be previously "for request header set A, we always got response header b after ACLs", but now "for request header set A, we can get either b1 or b2 depends on what upsteam returns", may be some decision has already been made ahead (like cached ACL result based on request hash?) and not prepared to be tampered. If this is adding request method as additional input rather than response I will feel more safe.

For the other code path, I can say it works and we have not observed unusual behavior in our daily use. But for this code path, I'm sorry, that is all I have now. I still not fully understand what that piece of code is used for, and why sending 100 Continue will trigger that path, and how will the modified reply be used. So I prefer to keep the original behavior of it.

I totally aggree that maintaining a good code structure and consistent logic is essential to a project. But this PR is an result in my company's daily work, and is already a rework from our original code, to make it similar to other parts. As long as I have more time, I should put more effort to keep our own projects in a good shape, rather than digging into a picece of code not related to our business.

No offensive, I just want to share this PR in case it is useful to community, don't know if this way is worse than not saying anything, or just paste something in newgroups, etc. And please take my apologize if I wasted community's time. Also please feel free to reject it if it not conforms to project standard.

Thanks.

Reply:
> it may be obvious to you that "there should be no problem"

It is not.

> For the other code path, I can say it works and we have not observed unusual behavior in our daily use

... but it can break previously "working" configurations just as changing the other code path may break other previously "working" configurations. It is great that the change works for you, but since we are modifying the official Squid version here, we need to consider more than just that. Please note that I am not saying that you have not considered it. I am just trying to explain why "working code" and "fear of change" are sometimes not good enough.

> I still not fully understand what that piece of code is used for,

It is used to forward an HTTP 1xx message (a.k.a "control message") from the origin server to the client.

> and why sending 100 Continue will trigger that path,

That path is specifically written to handle 1xx messages. 100 Continue is just a common example of such a message. I hope that example is handled by the same code paths that handles other 1xx control messages.

>  and how will the modified reply be used

It will be sent to the HTTP client, just like any other forwarded origin server reply. In some cases, it is important to police 1xx responses via reply_header_access and mangle them via httpHdrMangle(). YMMV.

> I just want to share this PR in case it is useful to community

You did the right thing. Your PR _is_ useful. Thank you! Hopefully, somebody will finish the work you have started to make these changes official.